openupgrade_scripts/scripts/loyalty/16.0.1.0/pre-migration.py

```diff
@@ -0,0 +1,630 @@
+    ("coupon.rule", "coupon_rule", "rule_min_quantity", "minimum_qty"),
+    ("coupon.rule", "coupon_rule", "rule_products_domain", "product_domain"),
+    (
+        "coupon.rule",
+        "coupon_rule",
+        "minimum_amount_tax_inclusion",
+        "minimum_amount_tax_mode",
+    ),
+]
+_models_renames = [
+    ("coupon.program", "loyalty.program"),
+    ("coupon.reward", "loyalty.reward"),
+    ("coupon.coupon", "loyalty.card"),
+    ("coupon.rule", "loyalty.rule"),
+]
+_tables_renames = [
+    ("coupon_program", "loyalty_program"),
+    ("coupon_reward", "loyalty_reward"),
+    ("coupon_coupon", "loyalty_card"),
+    ("coupon_rule", "loyalty_rule"),
+]
+_xmlids_renames = [
+    (
+        "coupon.coupon_action",
+        "loyalty.loyalty_card_action",
+    ),
+    (
+        "coupon.coupon_generate",
+        "loyalty.loyalty_generate_wizard_action",
+    ),
+    (
+        "coupon.report_coupon_code",
+        "loyalty.report_loyalty_card",
+    ),
+    (
+        "coupon.coupon_view_form",
+        "loyalty.loyalty_card_view_form",
+    ),
+    (
+        "coupon.coupon_view_tree",
+        "loyalty.loyalty_card_view_tree",
+    ),
+    (
+        "coupon.coupon_generate_view_form",
+        "loyalty.loyalty_generate_wizard_view_form",
+    ),
+    (
+        "coupon.coupon_program_view_form_common",
+        "loyalty.loyalty_program_view_form",
+    ),
+    (
+        "coupon.coupon_program_view_search",
+        "loyalty.loyalty_program_view_search",
+    ),
+    (
+        "coupon.coupon_program_view_tree",
+        "loyalty.loyalty_program_view_tree",
+    ),
+]
+
+
+def fill_date_to_field(env):
+    """
+    This data takes the values of the following fields:
+
+        validity_duration (coupon.program): in case of coupon programs that generate
+        coupons with a validity time, it is calculated by taking the creation date of
+        the program to which the duration defined in the field is added to obtain the
+        date.
+
+        rule_date_to (coupon.rule): in the case of promotional programs, the date of
+        rule_date_to is taken.
+    """
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE loyalty_program
+        ADD COLUMN IF NOT EXISTS date_to DATE
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE loyalty_program
+        SET date_to = CASE
+            WHEN validity_duration != 0 AND date_to IS NULL
+                THEN create_date + validity_duration * interval '1 day'
+            WHEN loyalty_rule.rule_date_to IS NOT NULL AND date_to IS NULL
+                THEN loyalty_rule.rule_date_to
+        END
+        FROM loyalty_rule
+        WHERE loyalty_program.validity_duration != 0
+        AND loyalty_program.rule_id = loyalty_rule.id
+        """,
+    )
+
+
+def fill_program_id_field(env):
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE loyalty_reward
```

Comment:
I think it would be better to group these changes by table

Reply:
done :)

Reply:
I still see a bunch of methods that could be group by table. I mean, a method for each table/model transformation

Reply:
done!